### DISCUSSION TITLE
What is the proper way to update store value based on props or state?

### DISCUSSION THREAD

**Discussion by baeharam** [mobxjs/mobx-state-tree#1957] [Q&A] [answered]
I'm using MST with React and read React new beta docs: [You Might Not Need an Effect](https://beta.reactjs.org/learn/you-might-not-need-an-effect). In the docs, it says that updating state based on state or props is bad idea. In this perspective, updating MST store value based on props or state with `useEffect` is bad idea too? Is there an alternative to replace this logic in MST? What is the proper way to update store value based on props or state?

Thanks for your reading :)

**Comment by EmilTholin** [comment, marked as answer]
Hi @baeharam!

If you want to update a MST store value based on props or state changes it might be worth taking a step back and evalute how you are interacting with your data. Do you have an example where you would like to update a MST store based on prop or state changes?

[A `useEffect` can be a nice way to sync prop changes into local component MobX state](https://mobx.js.org/react-integration.html#computed-props), but that's another problem entirely.

**Reply by baeharam** [reply]
Hello, @EmilTholin

For example, if I fetch some server data with MST store action and wanna update MST store value when that fetching is finished, do I have to use `useEffect` ?

**Reply by EmilTholin** [reply]
I would personally not use a `useEffect` in that case, but rather put the fetching logic and update logic right inside the `flow` action, and your `observer` React components will re-render automatically if they are using the MST model values in its rendering.

**Example**
```js
const MyModel = types
  .model("MyModel", {
    val: "",
  })
  .actions((self) => ({
    fetchVal: flow(function* () {
      const val = yield getVal();

      self.val = val;
    }),
  }));
```

I sometimes use `useEffect` to initiate a `flow` in a MST model, e.g. if your user is visiting the URL `/stuff` you might call an action in `useEffect` to fetch the `stuff` data.

**Example**
```js
const StuffPage = () => {
  const { stuffStore } = useStore();

  // Fetch stuff when the StuffPage component is first rendered
  useEffect(() => {
    stuffStore.fetchStuff();
  }, []);

  return <div>{/* Render stuff... */}</div>;
};

```

**Reply by baeharam** [reply]
Thanks for your answer!
